Add passphrase to ledger signing api

Yasuke opened the pull request:
This addresses #168; Instead of having the passphrase for signing
hardcoded to "Ledger.Crypto PassPhrase", allow the passphrase to be
provided. This way we aren't prevented from using private keys
created with tools like cardano-address and cardano-cli

silky reviewed (changes requested):
Looks pretty good @Yasuke , maybe I could suggest a few potential refactorings/additional items:

1. Perhaps introduce a new function, `signMessage`, signObservation'`, that use the default, and the un-ticked ones take the passphrase. This way we don't have random empty-strings in a few places,
2. Might not hurt to wrap up the PassPhrase in a newtype, so that it's a bit more clear it's purpose (and to allow later for nuanced handling, if we wish it),
3. Maybe we could take this opportunity to have a few tests that verify this behaviour? (Not sure what tests we have already around verifying signatures and passphrases?)

Thanks for the work so far! :)

volodyad commented:
Any progress on this?

volodyad commented:
Tried to build from PR branch - and it fails 
src/Cardano/Wallet/Mock/Handlers.hs:128:23: error:
    • Couldn't match expected type ‘Eff effs MockWallet’
                  with actual type ‘BS.ByteString -> Eff effs0 MockWallet’
    • Probable cause: ‘newWallet’ is applied to too few arguments

gege251 commented:
@silky @Yasuke sorry for urging you, but this issue is a blocker for other work, could we prioritize merging this once it builds?

silky commented:
Hi @gege251 ; sorry to hear it is blocking you. As you can see, we are working on it, and it will be merged in due course :)

Yasuke commented:
Thank you for the feedback and review @silky! I have make Passphrase a distinct type, added prime functions that don't require a password, added some tests (and instances needed to make them work) for signing in both the ledger and oracle.

Yasuke commented:
I also generalized the ledger signing function to work with any ByteArrayAccess instead of just TxIds

sjoerdvisscher commented on plutus-ledger/src/Ledger/CardanoWallet.hs:
```diff
@@ -68,10 +69,10 @@ newtype WalletNumber = WalletNumber { getWallet :: Integer }
 
 fromWalletNumber :: WalletNumber -> MockWallet
-fromWalletNumber (WalletNumber i) = fromSeed (BSL.toStrict $ serialise i)
+fromWalletNumber (WalletNumber i) = fromSeed (BSL.toStrict $ serialise i) ""
```
Use `noPassphrase` here.

Yasuke replied:
Adjusted, I used the ' version of the function that has no passphrase argument

sjoerdvisscher commented on plutus-use-cases/test/Spec/Future.hs:
```diff
@@ -203,7 +203,7 @@ settleEarly hdl = do
 
 mkSignedMessage :: POSIXTime -> Value -> SignedMessage (Observation Value)
-mkSignedMessage time vl = Oracle.signObservation time vl (fst oracleKeys)
+mkSignedMessage time vl = Oracle.signObservation time vl (fst oracleKeys) ""
```
```suggestion
mkSignedMessage time vl = Oracle.signObservation time vl (fst oracleKeys) noPassphrase
```

Yasuke replied:
Same with this one, I used signObservation' removing the passphrase argument

sjoerdvisscher commented on plutus-use-cases/test/Spec/Stablecoin.hs:
```diff
@@ -63,7 +63,7 @@ coin = Stablecoin
 
 signConversionRate :: POSIXTime -> ConversionRate -> SignedMessage (Observation ConversionRate)
-signConversionRate startTime rate = signObservation startTime rate oraclePrivateKey
+signConversionRate startTime rate = signObservation startTime rate oraclePrivateKey ""
```
```suggestion
signConversionRate startTime rate = signObservation startTime rate oraclePrivateKey noPassphrase
```

Yasuke replied:
Also here, signObservation'

sjoerdvisscher reviewed (approved):
Really nice work!

Yasuke commented:
Pending CI, required adjustments have been made, thanks @sjoerdvisscher for the review

silky commented on plutus-contract/src/Wallet/Emulator/Wallet.hs:
```diff
@@ -567,14 +567,14 @@ signTxWithPrivateKey :: (Member (Error WAPI.WalletAPIError) r) => PrivateKey ->
 signPrivateKeys :: [PrivateKey] -> SigningProcess
 signPrivateKeys signingKeys = SigningProcess $ \_ tx ->
-    pure (foldr Ledger.addSignature tx signingKeys)
+    pure (foldr (\pk ctx -> Ledger.addSignature' pk ctx) tx signingKeys)
```
```suggestion
    pure (foldr Ledger.addSignature' tx signingKeys)
```

Yasuke replied:
Good catch! fixed @silky

Yasuke commented:
Rebased on current main

volodyad commented:
hello, when this PR are planned to be merged?

sjoerdvisscher commented:
Thanks for the reminder!
